src/character.cpp

```diff
@@ -6006,6 +6006,115 @@ std::vector<body_part> Character::get_all_body_parts( bool only_main ) const
+    set_healthy( guy.healthy );
+    set_healthy_mod( guy.healthy_mod );
+    set_stim( guy.stim );
+    set_pain( guy.get_pain() );
+    set_painkiller( guy.pkill );
+    set_rad( guy.radiation );
+    str_max = guy.str_max;
+    dex_max = guy.dex_max;
+    int_max = guy.int_max;
+    per_max = guy.per_max;
+    str_bonus = guy.str_bonus;
+    dex_bonus = guy.dex_bonus;
+    per_bonus = guy.per_bonus;
+    int_bonus = guy.int_bonus;
+    reset_stats();
+    healthy = guy.healthy;
+    healthy_mod = guy.healthy_mod;
+    hunger = guy.hunger;
+    thirst = guy.thirst;
+    fatigue = guy.fatigue;
+    sleep_deprivation = guy.sleep_deprivation;
+    tank_plut = guy.tank_plut;
+    reactor_plut = guy.reactor_plut;
+    slow_rad = guy.slow_rad;
+    set_stamina( get_option<int>( "PLAYER_MAX_STAMINA" ) );
+    update_type_of_scent( true );
+    healthy_calories = guy.healthy_calories;
+    stored_calories = healthy_calories;
+    healed_total = guy.healed_total;
+    name = guy.name;
+    move_mode = CMM_WALK;
+    temp_cur = guy.temp_cur;
+    body_wetness = guy.body_wetness;
+    drench_capacity = guy.drench_capacity;
+    next_climate_control_check = guy.next_climate_control_check;
+    last_climate_control_ret = guy.last_climate_control_ret;
+    overmap_time = guy.overmap_time;
+    enchantment_cache = guy.enchantment_cache;
+    power_level = guy.power_level;
+    max_power_level = guy.max_power_level;
+    sight_max = guy.sight_max;
+    size_class = guy.size_class;
+    frostbite_timer = guy.frostbite_timer;
+    temp_conv = guy.temp_conv;
+    if( guy.mounted_creature ) {
+        guy.mounted_creature->mounted_player = this;
+        guy.mounted_creature->mounted_player_id = getID();
+        mounted_creature = std::move( guy.mounted_creature );
+    }
+    activity = player_activity();
+    cancel_activity();
+    clear_destination_activity();
+    stomach.set_nutr( guy.stomach.get_nutrients() );
+    stomach.set_water( guy.stomach.get_water() );
+    stomach.set_max_volume( guy.stomach.get_max_volume() );
+    stomach.set_contents( guy.stomach.get_contents() );
+    stomach.set_last_ate( guy.stomach.get_last_ate_point() );
+    guts.set_nutr( guy.guts.get_nutrients() );
+    guts.set_water( guy.guts.get_water() );
+    guts.set_max_volume( guy.guts.get_max_volume() );
+    guts.set_contents( guy.guts.get_contents() );
+    guts.set_last_ate( guy.guts.get_last_ate_point() );
+    oxygen = guy.oxygen;
+    consumption_history = guy.consumption_history;
+    focus_pool = guy.focus_pool;
+    clear_effects();
+    for( const bionic_id &bid : get_bionics() ) {
+        remove_bionic( bid );
+    }
+    for( const bionic_id &bid : guy.get_bionics() ) {
+        add_bionic( bid );
+    }
+    worn = guy.worn;
+    weapon = guy.weapon;
+    inv = guy.inv;
+    male = guy.male;
+    vitamin_levels = guy.vitamin_levels;
+    _skills = guy._skills;
+    player *pl = this->as_player();
+    if( pl ) {
+        pl->copy_from_npc_values_pl( guy );
+    }
```

Comment:
You can make `copy_from_npc_values` a virtual function of `Character` and `player` (and `creature` and `avatar`) instead of explicitly checking whether the `Character` is a `player` and calling `copy_from_npc_values_pl`.

Reply:
> You can make `copy_from_npc_values` a virtual function of `Character` and `player` (and `creature` and `avatar`) instead of explicitly checking whether the `Character` is a `player` and calling `copy_from_npc_values_pl`.

Would I then need to add the overrides to all subclasses of Creature? monster/avatar/player/npc/character, even when the function is only valid for player?

Seems a bit much

Reply:
You don't need to override it in all subclasses. If an override is not provided it just calls the function of the parent class.

And you don't have to make a general function. Override of each class just copies the value it needs, in addition to calling the function of the parent class, so the parent class also copies the value it needs.

Reply:
ah ok